Restore and support with solidus starter frontend

rahulsingh321 opened the pull request:
**Description:**
This PR aims at restoring the functionality of this extension with solidus latest version and provide support with solidus starter front-end.

**Following are the issues address in this PR -** 
1. #125 : Ensure compatibility with solidus starter front-end.
2. #124: Remove usage of initializer for configuring OAuth provider and use Spree::AuthenticationMethod table.
3. #115: Fixes loading issue of Spree::UserRegistrationController.
4. #85: Twitter OAuth support.
5. Fix issue related to missing routes
6. Fix for failing redirect after sign in.

rahulsingh321 committed:
Fix the admin authenticaton method form to show required fields

Fix stylesheet to use correct path to support starter FE

UI updates for the social logins

rahulsingh321 committed:
Update solidus social engine

- Implement ability to use omniauth from authentication method table not
from config
- Remove solidus_social initializer from generator and added devise as
initializer in extension
- Added validations for api_key and api_secet
- Fix for loading error for authentication method class

Fix for loading error of user registration controller

rahulsingh321 committed:
Fix for namespace and added required migrations

rahulsingh321 committed:
Fix for failing specs

rahulsingh321 committed:
Fixed the redirection issue on oauth

Align icons with starter frontend #125

Fix button styling

Minor fixes for social icons

Fix for loading and missing routes - #155

- Fix ::Spree::UserRegistrationsController loading error
- Fix for some missing routes

rahulsingh321 committed:
Fix for successful redirect path issue

Social auth buttons UI fixes

fthobe commented on config/routes.rb:
@rahulsingh321 hey, how come we need that line twice as spree_user_root and user_root?

rahulsingh321 replied:
Hi @fthobe This was the not proper solution, the issue here was root_path was not available in the extension, and devise redirect to the root_path after successfully signing. I override the `after_sign_in_path_for` and uses account_path as the path to use after sign in now.

fthobe commented on solidus_social.gemspec:
Please do not incute dev tools in PR

fthobe reviewed (changes requested):
Minor changes and some notes in general: 
- squash commits where possible, I see some smaller things like style changes that can be squashed
- Please be careful to separate what you work with (debug or developer tools) and what needs to be committed)
- circleci still gives an error, but I am sure that @kennyadsl or @jarednorman could take a look at that

kennyadsl commented on app/assets/stylesheets/spree/frontend/solidus_social.css:
```diff
@@ -1,4 +1,4 @@
 /*
- *= require spree/frontend
+ *= require spree/frontend/fontello
```
Could you explain this one, please?

rahulsingh321 replied:
Hi @kennyadsl I faced an issue after installing the extension i.e - somehow the sprockets is looking for the file `spree/frontend` to be there when compiling assets, and raised an error with Sprockets::FileNotFound. I have to change this so that it points to an file instead of frontend directory as Sprockets is unable to find. 

Here is the screenshot: 
![Screenshot-from-2024-12-24-21-56-42-12-24-2024_10_15_PM](https://github.com/user-attachments/assets/5e26c422-419a-45c3-bbad-0db0262c82d3)

Let me know if there any better way to handle this.

kennyadsl commented on lib/solidus_social.rb:
```diff
@@ -7,20 +7,37 @@
-  def self.configured_providers
-    ::Spree::SocialConfig.providers.keys.map(&:to_s)
-  end
+  OAUTH_PROVIDERS = [
```
Can you explain what's wrong with `configured_providers`? Why do we need to change it into a constant?

rahulsingh321 replied:
Hi @kennyadsl Earlier, we used to configure the providers using an initializer file, where we configure provider with credentials, I find it would be more appropriate if we pull the provider and credentials details from the `Spree::AuthenticationMethod` table instead, so I removed this method. 

Constant I took, just to maintain the list of provider into the extension itself. 
Let me know your thoughts on this.

kennyadsl replied:
To be honest I don't like to use the data from arrays stored in a constant. What about trying to save the information in a Struct or Data instance. Something like:

```
Provider = Data.define(:title, :key, :skip_signup)
Providers = [Project.new("GitHub", "github", true)]
```

This way we can use them in the codebase as `provider.key` instead of the current cryptic version: `provider[1]`.

fthobe replied:
@kennyadsl @rahulsingh321 I agree.

fthobe replied:
> @kennyadsl @rahulsingh321 I agree.

@kennyadsl please raise these kind of issues as soon as you can to keep the process of us restoring gems as streamlined as possible.

rahulsingh321 replied:
Thanks @kennyadsl Let me update this!

kennyadsl replied:
Thinking twice, I don't get why we are moving the credentials from the env variables to the database. Not saying it's wrong per-se, but I'd like to understand the reasons better and reading behind this choice, as it has security implications. Also, people currently using this extension with the env variables will have their store broken when upgrading, because their database doesn't contain the credentials.

rahulsingh321 replied:
@kennyadsl The decision to move credentials from environment variables to the database was made with scalability and automation in mind. Here's why I believe this change makes sense: 

1. Consistency in Configuration: It doesn't make sense for the admin to configure part of the authentication method from the backend (e.g., creating entries in the authentication methods table) while manually setting credentials in the environment variables. Centralising everything in the database makes for a more seamless and consistent configuration process.
2. Future Upgrades: I’ve found that the current implementation doesn’t support multi-store setups. With the database configuration, it’s much easier to extend this and connect it to the store. This way, each store can define its own authentication method with separate credentials, providing more flexibility for future upgrades.

3. Better for SaaS Products: For a SaaS product, using a database to store credentials is generally a better choice than relying on environment variables. It allows for better management, and scaling.

We need to update the readme here so that users can configure credentials accordingly while upgrading. I am not very sure what other best way to handle this more gracefully. 

Let me know what do you think on this.

kennyadsl replied:
Got it, thanks for the context. It makes sense but I guess we should provide a clear upgrade path. Considering people has those keys in their env already, what about writing them in the database along with the new migration?

rahulsingh321 replied:
@kennyadsl Done. Added the new migration to seed authentication method based one env variables.

kennyadsl commented:
Thanks for the contribution! Spec failures seem related to the changes, I don't think they are passing locally either.

rahulsingh321 committed:
Fix system specs and root_path issue

rahulsingh321 commented:
> Thanks for the contribution! Spec failures seem related to the changes, I don't think they are passing locally either.

Thank you @kennyadsl. I ran specs locally in SQLite and it worked, so circleCI. But, while running the specs with postgres/mysql it not creating the database automatically and we need this as we load the provider config from table now. Any thoughts on how to fix this ?

fthobe commented:
@kennyadsl @rahulsingh321 and also @shahmayur001 frequently struggle with testing on extension. Do you think it's possible to provide a better and more detailed example on how to implement testing for circleci in the docs? Literally something more step by step than what's present right now?

rahulsingh321 committed:
Added default scope for twitter for OAuth 2.0

rahulsingh321 committed:
Fix for initalizing all OAuth providers

fthobe commented:
@rahulsingh321 does this PR change in any way how return_url currently works or does it work as before (hence the cookie storing the url before the login event)?

[session[:spree_return_to].](https://legacy-guides.solidus.io/developers/customizations/customizing-after-login-redirects)

rahulsingh321 committed:
Revert user_id type change in spree_user_authentications for foreign key compatibility

rahulsingh321 committed:
Move OAuth providers initialization to devise.rb for proper loading

fthobe commented:
@rahulsingh321 can we wrapt it up together with this and close everything today with a final updated PR? 
#109

fthobe commented:
@kennyadsl all tests are passing on social. Can we merge this?

fthobe commented:
@rahulsingh321 we are ready to test today and wrap it up. 

Please take a look if we can add #109 and install the gem on staging, I will provide you with api tokens for the login methods later today.

fthobe commented:
@rahulsingh321 please also provide a proper PR description, I have updated the original issue #126 with all issues that are fixed by this PR.

kennyadsl commented:
@rahulsingh321 left a [comment](https://github.com/solidusio-contrib/solidus_social/pull/127#discussion_r1913378996) on the implementation. Let me know your thoughts.

rahulsingh321 committed:
Refactor: Use struct-based definition for OAUTH_PROVIDERS constant

kennyadsl commented on app/decorators/models/solidus_social/spree/user_decorator.rb:
```diff
@@ -12,9 +12,9 @@ def self.prepended(base)
-        skip_signup_providers = SolidusSocial::OAUTH_PROVIDERS.map { |p| p[1] if p[2] == 'true' }.compact
-        if skip_signup_providers.include? omniauth['provider']
-          self.email = omniauth['info']['email'] if email.blank?
+        skip_signup_providers = SolidusSocial::OAUTH_PROVIDERS.map { |p| p.key if p.skip_signup == true }.compact
```
`SolidusSocial::OAUTH_PROVIDERS.select { |provider| provider.skip_signup }.map(&:key)` is better for readability IMO.

kennyadsl reviewed (dismissed):
Left a non-blocking comment. In general, I'd try to avoid single-char variable names, which is not a best practice.

kennyadsl commented on lib/solidus_social/solidus_support.rb:
```diff
@@ -0,0 +1,9 @@
+  class << self
+    # Currently it checks only legacy frontend support
+    # enabling this here for starter frontend to load missing routes and controllers
+    def frontend_available?
```
I'm not sure about this. What's the purpose of the change? Will this change the value globally for all extensions? Why not change this in solidus support if this is a valid change?

fthobe replied:
@rahulsingh321 could you check this please?

rahulsingh321 replied:
Removed this and fixed the issue using route resolver.

rahulsingh321 committed:
Refactor skip_signup_providers with select

fthobe commented:
@rahulsingh321 this should work, please fix the tests

rahulsingh321 committed:
Fix missing route handling for starter frontend

- Added route resolver for explicit lookup from main_app for starter
frontend support.

fthobe commented:
@kennyadsl seems ok now

rahulsingh321 committed:
Added concurrent-ruby to < 1.3.5 for rails < 7.2

Issue:
https://stackoverflow.com/questions/79360526/uninitialized-constant-activesupportloggerthreadsafelevellogger-nameerror

rahulsingh321 committed:
Refactor route resolution to prioritize spree over main_app

rahulsingh321 committed:
Seed authentication methods with environment variables

rahulsingh321 committed:
Update README.md

fthobe commented:
@rahulsingh321 we will need to pin providers to stores in a future release.

rahulsingh321 commented:
> @rahulsingh321 we will need to pin providers to stores in a future release.

Sure @fthobe

fthobe commented:
@kennyadsl can we merge this? We have implemented all requested changes. To have this working with the installer also a minor Release of starter frontend needs to be made.

fthobe commented:
@kennyadsl depends on Starter Frontend release

kennyadsl commented on db/migrate/20250119155152_seed_authentication_methods_with_env_variables.rb:
```diff
@@ -0,0 +1,60 @@
+      api_secret: credentials[:api_secret]
+    ).destroy_all
+  end
+end
```
👏

kennyadsl commented on app/controllers/concerns/route_resolver.rb:
```diff
@@ -0,0 +1,12 @@
+module RouteResolver
```
Maybe this can be solved with something like: https://github.com/solidusio-contrib/solidus_reviews/pull/131/files#diff-14862ee5871d01a9196b6de5583b7b243fbf068a6c1e75e7db20e047c166358cR6? Did we already try?

rahulsingh321 replied:
Hi @kennyadsl Yes, I tried this already and this didn't help here. So, performed a explicit lookup the main_app to resolve missing routes.

kennyadsl reviewed (commented):
Thanks! I left a couple of notes but we are close!

tvdeyen commented on app/models/spree/authentication_method.rb:
```diff
@@ -1,19 +1,27 @@
-
-  validates :provider, presence: true
+class Spree::AuthenticationMethod < ActiveRecord::Base
+  validates :provider, :api_key, :api_secret, presence: true
```
Is this storing the credentials unencrypted in the database or do we have some mechanism to store them securely? If so, is this documented somewhere?

This should be addressed before we publish a new version to Rubygems. 

/cc @fthobe

tvdeyen commented:
@rahulsingh321 to ensure "compliance" in this gem with [OWASP](https://owasp.org/www-project-secure-coding-practices-quick-reference-guide/stable-en/01-introduction/05-introduction) we could use ActiveRecord's encrypted column feature.

Ideally we would use [Solidus build in preference store mechanisms](https://guides.solidus.io/advanced-solidus/model-preferences/) that allows to omit the database completely and store the secrets in the ENV. And in order to still support the feature to add the credentials in the admin we could use the `encrypted_string` preference value type.

/cc @fthobe @kennyadsl

fthobe commented:
> Ideally we would use Solidus build in preference store mechanisms that allows to omit the database completely and store the secrets in the ENV. And in order to still support the feature to add the credentials in the admin we could use the encrypted_string preference value type.

It makes sense to delay any development till it's clear what to do regarding devise.

tvdeyen commented:
> > Ideally we would use Solidus build in preference store mechanisms that allows to omit the database completely and store the secrets in the ENV. And in order to still support the feature to add the credentials in the admin we could use the encrypted_string preference value type.
> 
> It makes sense to delay any development till it's clear what to do regarding devise.

@fthobe this seems unrelated to this topic. Can you explain why?

fthobe commented:
Because it costs money and time and I don't believe reworking something here that currently "works" while we don't know where devise goes > maybe replacing it with a rails development that includes social as octa / azure became defacto standard for business and office / google / facebook covers pretty much the consumer side. 

I feel it's a waste of resources to fix something that works right now as long we don't know where auth is going. Doesn't mean I wouldn't like somebody else doing it :)

tvdeyen commented:
This issue introduced in this PR is not related to Devise at all. Have you read my comments and #138?
